atl-checker/src/algorithms/certain_zero/search_strategy/linear_optimize.rs

```diff
@@ -0,0 +1,515 @@
+use crate::game_structure::lcgs::ir::symbol_table::SymbolIdentifier;
+use crate::game_structure::lcgs::ast::{BinaryOpKind, DeclKind, ExprKind, Identifier};
+use crate::game_structure::lcgs::ir::intermediate::{IntermediateLCGS, State};
+use crate::edg::{Edge, ATLVertex};
+use crate::algorithms::certain_zero::search_strategy::{SearchStrategyBuilder, SearchStrategy};
+use minilp::{Problem, OptimizationDirection, ComparisonOp};
+use priority_queue::PriorityQueue;
+use std::collections::HashMap;
+use BinaryOpKind::{Addition,
+                   Multiplication,
+                   Subtraction,
+                   Division,
+                   Equality,
+                   Inequality,
+                   GreaterThan,
+                   LessThan,
+                   GreaterOrEqual,
+                   LessOrEqual,
+                   And,
+                   Or,
+                   Xor,
+                   Implication, };
+
+/// Holds extracted linear expressions from the formula in ATLVertex
+#[derive(Eq, PartialEq, Hash, Clone)]
+struct LinearExpression {
+    pub symbol: SymbolIdentifier,
+    pub constant: i32,
+    pub operation: BinaryOpKind,
+}
+
+/// Search strategy using ideas from linear programming to order next vertices,
+/// based on distance from the vertex to a region that borders the line between true/false in the formula.
+/// Has a cache to hold previously computed results, and uses a priority queue with
+/// distances as priority (lowest distance has highest priority)
+pub struct LinearOptimizeSearch {
+    queue: PriorityQueue<Edge<ATLVertex>, i32>,
+    game: IntermediateLCGS,
+    cache: HashMap<(Vec<LinearExpression>, usize), i32>,
+}
+
+impl LinearOptimizeSearch {
+    pub fn new(game: IntermediateLCGS) -> LinearOptimizeSearch {
+        LinearOptimizeSearch {
+            queue: PriorityQueue::new(),
+            game,
+            cache: HashMap::new(),
+        }
+    }
+}
+
+/// A SearchStrategyBuilder for building the LinearOptimizeSearch strategy.
+pub struct LinearOptimizeSearchBuilder {
+    pub game: IntermediateLCGS,
+}
+
+impl SearchStrategyBuilder<ATLVertex, LinearOptimizeSearch> for LinearOptimizeSearchBuilder {
+    fn build(&self) -> LinearOptimizeSearch {
+        LinearOptimizeSearch::new(self.game.clone())
+    }
+}
+
+impl SearchStrategy<ATLVertex> for LinearOptimizeSearch {
+    /// Simply returns the edge with highest priority (i.e lowest distance)
+    fn next(&mut self) -> Option<Edge<ATLVertex>> {
+        let edge = self.queue.pop();
+        if edge.is_some() {
+            Some(edge.unwrap().0)
+        } else { None }
```

Comment:
This whole block can be replaced with
```rust
self.queue.pop().map(|entry| entry.0)
```

Reply:
fixed in 86f10b70fa28dd237a42c4d12482cdf6672b737c